ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java

```diff
@@ -0,0 +1,554 @@
+    public byte[] serialize(String topicName, Object spec) {
+      String schemaString;
+      try {
+        schemaString = schemaRegistryClient.getLatestSchemaMetadata(topicName + "-value").getSchema();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      new org.apache.avro.Schema.Parser().parse(schemaString);
+      Object avroObject = valueSpecToAvro(
+          spec,
+          new org.apache.avro.Schema.Parser().parse(schemaString));
+      return avroSerializer.serialize(topicName, avroObject);
+    }
+  }
+
+  protected static class ValueSpecAvroSerdeSupplier implements SerdeSupplier<Object> {
+    @Override
+    public Serializer<Object> getSerializer(SchemaRegistryClient schemaRegistryClient) {
+      return new ValueSpecAvroSerializer(schemaRegistryClient);
+    }
+
+    @Override
+    public Deserializer<Object> getDeserializer(SchemaRegistryClient schemaRegistryClient) {
+      return new ValueSpecAvroDeserializer(schemaRegistryClient);
+    }
+  }
+
+  protected static class ValueSpecJsonDeserializer implements Deserializer<Object> {
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(Map<String, ?> properties, boolean b) {
+    }
+
+    @Override
+    public Object deserialize(String topicName, byte[] data) {
+      try {
+        return new ObjectMapper().readValue(data, Map.class);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  protected static class ValueSpecJsonSerializer implements Serializer<Object> {
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(Map<String, ?> properties, boolean b) {
+    }
+
+    @Override
+    public byte[] serialize(String topicName, Object spec) {
+      try {
+        return new ObjectMapper().writeValueAsBytes(spec);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  protected static class ValueSpecJsonSerdeSupplier implements SerdeSupplier<Object> {
+    @Override
+    public Serializer<Object> getSerializer(SchemaRegistryClient schemaRegistryClient) {
+      return new ValueSpecJsonSerializer();
+    }
+
+    @Override
+    public Deserializer<Object> getDeserializer(SchemaRegistryClient schemaRegistryClient) {
+      return new ValueSpecJsonDeserializer();
+    }
+  }
+
+  protected static class Topic {
+    private final String name;
+    private final String format;
+    private final org.apache.avro.Schema schema;
+
+    Topic(final String name, final String format, final org.apache.avro.Schema schema) {
+      this.name = name;
+      this.format = format;
+      this.schema = schema;
+    }
+
+    public String getName() {
+      return name;
+    }
+
+    public String getFormat() {
+      return format;
+    }
+
+    public org.apache.avro.Schema getSchema() {
+      return schema;
+    }
+  }
+
+  protected static class Record {
+    private final String topic;
```

Comment:
Wondering if this should be `Topic` now? and the `SerdeSupplier` could go on the `Topic`?

Reply:
Yes this makes sense.